feat(sdk-analytics): add the ability to send custom events to Analytics

oidacra opened the pull request:
## Summary

  Major refactoring of the Analytics SDK to improve architecture, simplify payload structure, and enhance the event tracking system with clearer separation of concerns between
  automatic enrichment and user-provided data.

  ## Changes Made

  ### Core Architecture

  - **Payload Structure Refactoring** - Reorganized event payload handling to separate automatic enrichment (page, device, UTM data) from user-provided custom properties
  - **Model Organization** - Split monolithic model file into focused, domain-specific model files:
    - `data.model.ts` - Page, device, and UTM data structures
    - `event.model.ts` - Event-specific types and request bodies
    - `library.model.ts` - Public API types (reduced from 300+ to ~70 lines)
    - `request.model.ts` - HTTP request/response types
  - **Plugin Simplification** - Streamlined enricher and analytics plugins by moving enrichment logic upstream, reducing plugin complexity by ~60%

  ### Event Handling

  - **Enricher Plugin** (`dot-analytics.enricher.plugin.ts:65-129`)
    - Now builds complete request body structure with `context` and `events` arrays
    - Separates user properties into `custom` data object for clarity
    - Prevents reserved property pollution (filters out Analytics.js internal properties)
    - Validates event names (prevents "pageview" as custom event name)
  - **Analytics Plugin** (`dot-analytics.plugin.ts:30-70`)
    - Simplified to pass-through fully-enriched payloads from enricher
    - Removed redundant payload restructuring logic
    - Cleaner separation: enricher builds, analytics sends

  ### Data Models

  - **New `custom` Property Pattern** - User-provided properties now wrapped in `custom` object to distinguish from automatic enrichment
  - **Request Body Consolidation** - Unified pageview and track event structures under consistent `events` array format
  - **Type Safety Improvements** - Added `DotCMSAnalyticsEnrichedParams` and `DotCMSEventData` for better compile-time checks

  ### Documentation

  - **README Overhaul** (`README.md:32-387`)
    - Complete rewrite with clearer API documentation
    - Added detailed payload structure examples showing automatic vs. custom data
    - Improved usage examples for vanilla JS and React
    - Added "Under the Hood" section explaining enrichment flow
    - Documented all automatic data capture (page, device, UTM, context)
    - Clarified that React integration is still in development

  ### Testing

  - **Test Coverage Expansion** - Updated all test suites to match new payload structure:
    - `dot-analytics.enricher.plugin.spec.ts` - Tests for new `custom` data wrapping
    - `dot-content-analytics.http.spec.ts` - Simplified to match unified request body format
    - `useContentAnalytics.spec.tsx` - React hook tests updated for new API

  ### Utility Functions

  - **Property Filtering** (`dot-analytics.enricher.plugin.ts:16-25`)
    - Added `INTERNAL_ANALYTICS_KEYS` constant to filter Analytics.js metadata
    - Cleans user properties before wrapping in `custom` object
  - **Session Management** - Moved session activity updates to identity plugin for consistency

  ## Technical Details

  ### Key Architectural Changes

  1. **Three-Stage Pipeline**: Identity Plugin (context) → Enricher Plugin (data + structure) → Analytics Plugin (send)
  2. **Property Namespacing**: Automatic enrichment at root level, user data under `custom`
  3. **Request Body Standardization**: Both pageview and custom events use identical `{ context, events[] }` structure

  ### Event Payload Flow

  ```typescript
  // User calls: track('button-click', { label: 'CTA' })
  ↓
  // Enricher builds:
  {
    context: { site_key, session_id, user_id },
    events: [{
      event_type: 'button-click',
      local_time: '2024-01-01T10:00:00Z',
      data: { custom: { label: 'CTA' } }
    }]
  }
  ↓
  // Analytics sends to server

```

  ### Breaking Changes

  Payload Structure Changes (Internal Only - No Public API Changes):
  - User properties now wrapped in data.custom instead of data root
  - Enriched payload structure changed from flat object to structured { context, events[] } format

  Note: These are internal implementation changes. The public API (track() and pageView() methods) remains unchanged.

  ### Testing

  - Unit tests added/updated - 28 test files updated
  - Integration tests verified - HTTP request body structure validated
  - Manual testing performed - Requires verification in demo environment
  - E2E tests added/updated - N/A for SDK changes

  ### Related Issues

  Closes #33396

 ### Additional Notes

  - React Integration Status: Marked as "In Development" in documentation to set proper expectations
  - Migration Path: Internal payload changes are transparent to SDK consumers

### Screenshots

https://github.com/user-attachments/assets/f2d85e14-81dc-456a-a494-01f444e241f3

### How to use
https://github.com/dotCMS/core/blob/main/examples/nextjs/src/components/forms/ContactUs.js

```
import { analyticsConfig } from "@/utils/analyticsConfig"; // this is the centralized config for analytics
import { useContentAnalytics } from "@dotcms/analytics/react";  // import from analytics the main custom hook

...

export default function ContactUs({ description }) {
    ...

    const { track } = useContentAnalytics(analyticsConfig); // get the track function from the custom hook

    ...

    // Track the form submission only when the form is submitted
    const handleTrack = () => {
        track("form-submit", { formName: "contact-form", formType: "lead-gen" });
    }

    // Emulate a form submission
    const handleSubmit = (e) => {
        e.preventDefault();
        setIsSubmitting(true);
        setTimeout(() => {
            setIsSuccess(true);
            handleTrack();  // call the track handler
            resetForm();
        }, 3000);
    };

    return (
        <div className="max-w-lg mx-auto my-10 p-6 bg-white rounded-lg shadow-md">
            <h2 className="text-2xl font-bold text-gray-800 mb-6">
                Contact Us
            </h2>

                   ...
                    <button
                       type="submit"
                        disabled={isSubmitting}
                        className="px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:opacity-75"
                    >
                        {isSubmitting ? "Submitting..." : "Submit"}
                    </button>
                </div>
            </form>
        </div>
    );
}

```

This PR fixes: #33396

This PR fixes: #33396

oidacra committed:
refactor(analytics): streamline event handling and payload enrichment

### Summary of Changes
- Updated the  and  to utilize enriched payloads directly, simplifying the event tracking process.
- Removed redundant checks and streamlined the structure of the request bodies for page view and custom events.
- Enhanced type definitions for analytics parameters to clearly differentiate between raw and enriched payloads.
- Added debug logging for outgoing HTTP requests to aid in troubleshooting.

oidacra committed:
refactor(analytics): reorganize constants and models for clarity

- Refactored imports to utilize a centralized  and  structure, improving code organization.
- Updated references across various files to align with the new structure, ensuring consistency in the analytics SDK.
- Introduced new model definitions for analytics data and events, enhancing type safety and clarity in the codebase.

oidacra committed:
refactor(analytics): enhance documentation and improve event tracking structure

- Updated the README.md to clarify the React integration status and improve the API documentation for page views and custom events.
- Enhanced the structure of the event payloads to include automatic enrichment details, making it clearer how data is captured.
- Refactored test cases to remove redundant session activity updates and improve clarity in event tracking assertions.
- Consolidated constants and model imports for better organization and maintainability.

oidacra committed:
refactor(analytics): improve user-provided properties filtering

oidacra committed:
refactor(sdk-analytics): clean up test file formatting and imports

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/plugin/identity/dot-analytics.identity.plugin.ts:
```diff
@@ -38,8 +36,10 @@ export const dotAnalyticsIdentityPlugin = (config: DotCMSAnalyticsConfig) => {
         /**
          * Inject identity context into page events
          * This runs BEFORE the enricher plugin
+         * Also updates session activity for session management
```
```suggestion
        * Inject identity context into page events and updates session activity for session management
        * This runs BEFORE the enricher plugin
```

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/plugin/identity/dot-analytics.identity.plugin.ts:
```diff
@@ -51,8 +51,10 @@ export const dotAnalyticsIdentityPlugin = (config: DotCMSAnalyticsConfig) => {
         /**
          * Inject identity context into track events
          * This runs BEFORE the enricher plugin
+         * Also updates session activity for session management
```
```suggestion
        * Inject identity context into track events and updates session activity for session management
        * This runs BEFORE the enricher plugin
```

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/plugin/dot-analytics.plugin.ts:
```diff
@@ -38,94 +30,41 @@ export const dotAnalytics = (config: DotCMSAnalyticsConfig) => {
         /**
          * Track a custom event
-         * Takes enriched data and sends it to the analytics server
+         * The enricher plugin has already built the complete request body
```
Kind of smell that we have to document that one plugin do one thing... so if we don't use the enricher plugins, the whole thing breaks?

oidacra replied:
We can't use analytics without an enricher.
It is instantiated directly with all the plugins loaded.
```
    const analytics = Analytics({
        app: 'dotAnalytics',
        debug: config.debug,
        plugins: [
            dotAnalyticsIdentityPlugin(config), // Inject identity context (user_id, session_id, local_tz)
            dotAnalyticsEnricherPlugin(), // Enrich and clean payload with page, device, utm data and custom data
            dotAnalytics(config) // Send events to server
        ]
    });
```

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/constants/dot-content-analytics.constants.ts:
```diff
@@ -8,14 +8,30 @@ export const ANALYTICS_SOURCE_TYPE = ANALYTICS_WINDOWS_KEY;
+ * Custom event type - any string except predefined event types
+ * These events have flexible data structures defined by the user
+ */
+export type CustomEventType = Exclude<string, DotCMSPredefinedEventType>;
```
```suggestion
export type DotCMSCustomEventType = Exclude<string, DotCMSPredefinedEventType>;
```

oidacra replied:
done

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/data.model.ts:
```diff
@@ -0,0 +1,80 @@
+ */
+export interface DotCMSDeviceData {
+    /** Screen resolution as a string (e.g., "1920x1080") */
+    screen_resolution: string | undefined;
+    /** User's preferred language */
+    language: string | undefined;
+    /** Viewport width in pixels */
+    viewport_width: string | undefined;
+    /** Viewport height in pixels */
+    viewport_height: string | undefined;
```
When all of this can be optional?

oidacra replied:
After a detailed review. This can never be optional.

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/data.model.ts:
```diff
@@ -0,0 +1,80 @@
+ */
+export interface DotCMSPageData {
+    /** The current page URL */
+    url: string | undefined;
+    /** Document encoding */
+    doc_encoding: string | undefined;
+    /** Document hash fragment */
+    doc_hash: string;
+    /** Document protocol (http/https) */
+    doc_protocol: string | undefined;
+    /** Document search parameters */
+    doc_search: string;
+    /** Document host domain */
+    doc_host: string | undefined;
+    /** Document path */
+    doc_path: string | undefined;
+    /** Page title */
+    title: string | undefined;
```
When all of this can be optionals?

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/event.model.ts:
```diff
@@ -0,0 +1,61 @@
+        /** UTM parameters for campaign tracking (optional) */
+        utm?: DotCMSUtmData;
+        /** Custom data associated with the event */
+        custom?: Record<string, unknown>;
```
Does pageview can have custom props? and if so... why id not `data.custom` like the `DotCMSCustomEvent`?

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/event.model.ts:
```diff
@@ -0,0 +1,61 @@
+ */
+export interface DotCMSCustomEvent extends DotCMSEventBase {
+    /** The type of event being tracked (any string except structured events) */
+    event_type: CustomEventType;
```
```suggestion
    event_type: event_type: typeof DotCMSPredefinedEventType.CUSTOM;;
```

oidacra replied:
the last one
```
event_type: typeof DotCMSPredefinedEventType.CUSTOM;;
```

The event_type is whatever the client wants, is any string except `pageview` (by now, I will add later content_impression and so on.)

```
export type DotCMSCustomEventType = Exclude<string, DotCMSPredefinedEventType>;
```

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/event.model.ts:
```diff
@@ -0,0 +1,61 @@
+    /** The type of event being tracked (any string except structured events) */
+    event_type: CustomEventType;
+    /** Custom event data with structured format */
+    data: {
```
Why here is `data.custom` but in the other one is just `custom`?

fmontes commented on core-web/libs/sdk/analytics/src/lib/core/shared/models/library.model.ts:
Why this is not typed?

fmontes commented on core-web/libs/sdk/analytics/src/lib/react/hook/useContentAnalytics.ts:
```diff
@@ -1,33 +1,44 @@
  * Custom hook that handles analytics tracking for anonymous users.
- * Provides methods to track events and page views with automatic timestamp injection.
- * Automatically disables tracking when inside the UVE editor.
+ * Provides methods to track events and page views.
+ *
+ * **UVE Editor Behavior:**
+ * - Automatically disables ALL tracking when inside the Universal Visual Editor (UVE)
+ * - This prevents editor interactions and preview activities from polluting analytics data
+ * - UVE detection is memoized for performance - checked once per component lifecycle
+ *
+ * **Performance:**
+ * - Uses a singleton pattern - all components with the same config share the same analytics instance
+ * - Only re-initializes when server or siteKey changes (not debug flag)
+ * - UVE state check is memoized to avoid repeated function calls
```
This is a doc that our users will read; it should tell them how to use this, not explain how we built it.

oidacra committed:
refactor(sdk-analytics): standardize logging messages and improve error handling

- Updated console warning and error messages to consistently use "DotCMS Analytics" prefix for better clarity.
- Refactored various analytics functions to enhance error handling and improve the overall structure of the codebase.
- Adjusted type definitions and payload structures to align with the new naming conventions and improve type safety.
- Enhanced documentation for better understanding of the analytics SDK functionality.

oidacra committed:
refactor(analytics): update UTM parameter extraction and improve test coverage

- Refactored the `extractUTMParameters` function to accept `window.location` as a parameter for better flexibility.
- Updated related test cases to reflect the new function signature and ensure comprehensive coverage of UTM parameter extraction.
- Enhanced documentation for clarity on UTM parameter handling and its significance in campaign tracking.

oidacra committed:
refactor(analytics): enhance type safety and improve payload structures

### Summary of Changes
- Introduced `JsonObject` and `JsonArray` types to standardize the structure of custom data across analytics events.
- Updated function signatures for `pageView` and `track` methods to use the new `JsonObject` type, improving type safety.
- Refactored event models to include structured data types for page view and custom events, ensuring consistency in analytics payloads.
- Enhanced documentation to clarify the expected data formats for analytics tracking.

oidacra committed:
refactor(sdk-analytics): enhance event type definitions and improve payload structure

oidacra committed:
refactor(sdk-analytics): rename site_key to site_auth in DotCMSAnalyticsEventContext

oidacra committed:
refactor(sdk-analytics): update test cases to reflect siteAuth changes

oidacra committed:
Merge branch 'main' into 33396-add-custom-events-sdk-analytics
